src/ir/effects.h

```diff
@@ -701,6 +701,10 @@ class EffectAnalyzer {
       }
     }
     void visitCallRef(CallRef* curr) {
+      if (curr->target->type.isNull()) {
+        parent.trap = true;
```

Comment:
Are you thinking that we should be setting `implicitTrap` when the target is null rather than setting `trap`? I think it's more consistent to set `trap`, just like `visitUnreachable` does. In particular, the `CallRef` will be emitted as `(unreachable)` in this case, so it would be strange if its effects were different from `Unreachable`'s effects.

You're probably right that optimizations would take care of this anyway, but the code seems more correct to me as-is.

Reply:
I was thinking we can set `implicitTrap` when the target is nullable, which would include null. You're right that the current code is more precise in a way, but I don't think we gain from the extra verbosity.

I guess it's fine, though, if we make sure that it doesn't limit us. Specifically, with trapsNeverHappen and the new code then a callRef of a null would no longer be removed by vacuum (like it doesn't remove an unreachable). We should have a test that optimize-instructions turns the callRef into an unreachable (which dce would then remove). (There might already be a test, sorry if so - I haven't gotten to those yet.)

Reply:
It seems odd that `trapsNeverHappen` doesn't remove an `unreachable`, but I guess that's a separate topic.

Will make sure I add specific tests for the OptimizeInstructions changes.